datafusion/expr/src/function.rs

```diff
@@ -38,9 +39,20 @@ pub type ReturnTypeFunction =
     Arc<dyn Fn(&[DataType]) -> Result<Arc<DataType>> + Send + Sync>;
 
 /// Factory that returns an accumulator for the given aggregate, given
-/// its return datatype.
-pub type AccumulatorFactoryFunction =
-    Arc<dyn Fn(&DataType) -> Result<Box<dyn Accumulator>> + Send + Sync>;
+/// its return datatype, the sorting expressions and the schema for ordering.
+pub type AccumulatorFactoryFunction = Arc<
+    dyn Fn(&DataType, &[Expr], &Schema) -> Result<Box<dyn Accumulator>> + Send + Sync,
+>;
+
+/// Factory that returns an accumulator for the given aggregate, given
+/// its return datatype, the sorting expressions and the schema for ordering.
+/// FirstValue needs additional `ignore_nulls` and `requirement_satisfied` flags.
+// TODO: It would be nice if we can have flexible design for arbitrary arguments.
```

Comment:
Before we have a flexible design, I plan to have one default function that has the common arguments and specifically define others that have additional arguments. Like what it is now.

Reply:
It might also be useful to add some way for the accumulator to specify what options it permits so DataFusion can do the error checking

Maybe

```rust
struct AccumulatorOptions {
  /// Aggregate supports `ORDER BY` to specify the order of arguments
  pub bool sort_exprs,
  /// Aggregate supports `IGNORE NULLS` to specify null treatment
  pub bool ignore_nulls,
...
}
```

That way the UDAFs could opt in to supporting this kind of syntax

Reply:
Nice idea! 😄